stan/math/prim/prob/discrete_range_rng.hpp

```diff
@@ -0,0 +1,65 @@
+#ifndef STAN_MATH_PRIM_PROB_DISCRETE_RANGE_RNG_HPP
+#define STAN_MATH_PRIM_PROB_DISCRETE_RANGE_RNG_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/max_size.hpp>
+#include <boost/random/uniform_int_distribution.hpp>
+#include <boost/random/variate_generator.hpp>
+
+namespace stan {
+namespace math {
+
+/** \ingroup prob_dists
+ * Return a discrete random variate for the given lower and upper bounds
+ * (inclusive) using the specified random number generator.
+ *
+ * `lower` and `upper` can each be a scalar or a one-dimensional container.
+ * Any non-scalar inputs must be the same size.
+ *
+ * @tparam T_lower type of lower bound
+ * @tparam T_upper type of upper bound
+ * @tparam RNG type of random number generator
+ *
+ * @param lower lower bound
+ * @param upper upper bound
+ * @param rng random number generator
+ * @return A (sequence of) discrete random variate(s) between `lower` and
+ * `upper`, both bounds included.
+ * @throw std::domain_error if lower or upper are non-finite, or if upper
+ * is smaller than lower
+ * @throw std::invalid_argument if non-scalar arguments are of different
+ * sizes
+ */
+template <typename T_lower, typename T_upper, class RNG>
+inline typename VectorBuilder<true, int, T_lower, T_upper>::type
+discrete_range_rng(const T_lower& lower, const T_upper& upper, RNG& rng) {
```

Comment:
The documentation needs to indicate what `T_lower` and `T_upper` can be (presumably either `int` or `std::vector<int>`).  

It should not be possible to compile an instantiation where the lower or upper bounds are `double`.  It's OK if that's enforced by not being able to compile someting inside---it's not like we have to guard against ambiguity.

Reply:
I've updated the docs. See my other answer as for the instantiation with non-int arguments.